Fcc detector migration

atolosadelgado opened the pull request:
BEGINRELEASENOTES

* Migration of IDEA and ALLEGRO detectors from FCCDetectors repository, becoming each one the option 1 version 01 of the corresponding detector
	- The compact files and detector constructors files, and the detector name therein, are renamed to avoid collision with FCCDetector components by appending the suffix `_o1_v01`
	- IDEA compact files are copied into `./FCCee/IDEA/compact/IDEA_o1_v01`
	- ALLEGRO compact files are copied into `./FCCee/ALLEGRO/compact/ALLEGRO_o1_v01`
	- The detector constructors are placed in the corresponding directory inside `./detector`, and therein in the appropriate sub-directory;
		- `calorimeter`: CaloEndcapDiscs_o1_v01_geo.cpp, HCalThreePartsEndcap_o1_v01_geo.cpp, HCalTileBarrel_o1_v01_geo.cpp
		- `other`: SimpleCylinder_geo_o1_v01.cpp
		- `tracker`: parametrised_DriftChamber_o1_v01.cpp
	- DD4hep segmentations from FCCDetectors were also migrated into the directory `detectorSegmentations`, and renamed by appending the suffix `_k4geo`

ENDRELEASENOTES

Hi all, 

this PR implements the changes stated here 
https://github.com/key4hep/k4geo/issues/267#issue-1712105098

this PR is not ready yet, and some things are open to discussion. Particularly, how to handle the non-detector parts migrated from FCC detector repo. At the moment they are placed in detector/{DetSegmentation, DetCommon, DetSensitive}

DetCommon defines a number of global functions that are needed mainly to reconstruction. Only the function cellID implemented in DetCommon is used in FCCDetectors by DetSensitive.

Thank you for your time.
Alvaro

andresailer commented:
As discussed in the Key4hep meeting, it would be better to remove things that are not needed, for example in the DetCommon folder.

BrieucF commented on detector/calorimeter/ECalBarrelInclined_geo.cpp:
This one was migrated in this PR: https://github.com/key4hep/k4geo/pull/276

BrieucF commented:
Breaking news: you can change `NLB` to `ALLEGRO` !

atolosadelgado commented:
Hi, 

I purged some files out of DetCommon. I kept for testing purposes a simple box dummy detector. The simple box was used to test the Sensitive Detector actions, but actually the cmake test command is now commented. I still think a test is healthy for the repo, but since this PR is urgent I did not spend time on rewriting the test. 

I moved the segmentation and senstive detector folders to main folder and renamed according to the proposal in https://github.com/key4hep/k4geo/issues/267

a test Gaudi simulation was run to check if the changes were consistent.

atolosadelgado commented:
Hi, 

as discussed in the key4hep meeting, i have removed the DetCommon and sensitive detectors folder. The segmentation classes have been renamed, adding `_k4geo` to the name. 

The detectors from FCCDetectors repo are also migrated, xml and cpp files. The name is also changed so they will not clash.

I hope there are not many issues... I changed a lot of code

Alvaro

BrieucF commented:
Any further comment on this @andresailer ?

andresailer commented on detector/calorimeter/CaloEndcapDiscs_geo.cpp:
Can you add a few phrases about what this detector looks like?

andresailer commented on detector/calorimeter/HCalBarrel_geo.cpp:
Can you add a few phrases about what this detector looks like?

andresailer commented on detector/calorimeter/HCalThreePartsEndcap_geo.cpp:
Can you add a few phrases about what this detector looks like?

andresailer commented on detectorSegmentations/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+
+file(GLOB headers include/detectorSegmentations/*.h)
+set_target_properties(detectorSegmentations PROPERTIES PUBLIC_HEADER "${headers}")
+set_property(TARGET detectorSegmentations PROPERTY GLOBAL)
```
Does this do anything? I don't see that GLOBAL is a property for targets:
https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#properties-on-targets

andresailer commented on detectorSegmentations/include/detectorSegmentations/kk.txt:
What is this file?

andresailer commented on detector/tracker/parametrised_DriftChamber.cpp:
Can you add a few phrases about this detector?

atolosadelgado commented on CMakeLists.txt:
```diff
@@ -80,7 +81,7 @@ target_include_directories(${PackageName}G4 PRIVATE ${PROJECT_SOURCE_DIR}/detect
 target_include_directories(${PackageName}G4 SYSTEM PUBLIC ${LCIO_INCLUDE_DIRS})
 
-target_link_libraries(${PackageName}   DD4hep::DDCore DD4hep::DDRec DD4hep::DDParsers ${ROOT_LIBRARIES} ${LCIO_LIBRARIES})
+target_link_libraries(${PackageName}   DD4hep::DDCore DD4hep::DDRec DD4hep::DDParsers ${ROOT_LIBRARIES} ${LCIO_LIBRARIES} detectorSegmentations)
```
The file `detector/tracker/parametrised_DriftChamber.cpp` requires a segmentation class: `detectorSegmentations/GridSimplifiedDriftChamber_k4geo.h` 

When I tried to include just the header instead of linking, like:
```
target_include_directories(${PackageName}   PRIVATE ${PROJECT_SOURCE_DIR}/detectorSegmentations/include )

target_link_libraries(${PackageName}   DD4hep::DDCore DD4hep::DDRec DD4hep::DDParsers ${ROOT_LIBRARIES} ${LCIO_LIBRARIES} )

```
, it raises an error at the end of the compilation,
```
/cvmfs/sw.hsf.org/contrib/binutils/2.38/x86_64-centos7-gcc8.3.0-opt/pv4x2/bin/ld: CMakeFiles/lcgeo.dir/detector/tracker/parametrised_DriftChamber.cpp.o: in function `det::createParamSimplifiedDriftChamber(dd4hep::Detector&, dd4hep::xml::Handle_t, dd4hep::SensitiveDetector)':
parametrised_DriftChamber.cpp:(.text+0x466): undefined reference to `typeinfo for dd4hep::DDSegmentation::GridSimplifiedDriftChamber_k4geo'
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/lcgeo.dir/build.make:2057: lib/liblcgeo.so] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:158: CMakeFiles/lcgeo.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

andresailer replied:
Ok, then it is fine to leave it in.

atolosadelgado commented:
I think I have documented the corresponding READMEs
I still do not know what to do with the dependency of the drift chamber on segmentation library.

andresailer commented on detector/calorimeter/HCalBarrel_geo.cpp:
Rename to HCalTileBarrel_o1_v01_geo.cpp

atolosadelgado commented:
Should some release note comments be filled in the OP?

andresailer commented:
> Should some release note comments be filled in the OP?

Yes, definitely! Thanks!

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    // second z loop (place sequences in layer)
+    std::vector<dd4hep::PlacedVolume> sequences; 
```
There is a compiler warning here:
```
 /Package/detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:201:39: warning: declaration of 'sequences' shadows a previous local [-Wshadow]
  201 |     std::vector<dd4hep::PlacedVolume> sequences;
      |                                       ^~~~~~~~~
/Package/detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:50:27: note: shadowed declaration is here
   50 |   std::vector<xml_comp_t> sequences = {xmlDet.child(_Unicode(sequence_a)), xmlDet.child(_Unicode(sequence_b))};
      |                           ^~~~~~~~~
```

atolosadelgado commented:
Hi,

I have displayed the geometries of IDEA and ALLEGRO with geoDisplay tool, it worked. No warnings during compilation.

I found that for some reason the Drift chamber and calorimeter of IDEA were commented out from the main file, but now they are ready. 

Anyway, these components were templates for the future ones, some of them already in PR or upcoming soon.

I can not think of further refinements, but please let me know if I can do something else.
